### ISSUE TITLE
How to toggle the 'abort_on_first_run_crash'-option

### ISSUE THREAD

**Issue by LishengSun** [automl/auto-sklearn#345]
Hi Matthias,

I am using auto-sklearn (2.0), and it quit when the first run crashes: 
`smac.tae.execute_ta_run.FirstRunCrashedException: First run crashed, abort. (To prevent this, toggle the 'abort_on_first_run_crash'-option!)`

Can you give more instructions on how to toggle the 'abort_on_first_run_crash'-option? Thank you very much!

**Comment by mlindauer** [comment]
Hi,

since Matthias is away and I don't know when he can answer, let me quickly answer.
Unfortunately, this option is hidden inside of SMAC and not transparent for AutoSklearn users.

First of all, please let me note that in most cases something is broken with your setup if the first run crashes. So, you should first ensure that there is no bug involved triggering this crashed run.

If you really want to avoid this Exception, you could go to the following code:
https://github.com/automl/auto-sklearn/blob/master/autosklearn/smbo.py#L516

If you add to the `scenario_dict`, 
```"abort_on_first_run_crash": False```
this error should not be raised.

Best,
Marius

**Comment by LishengSun** [comment]
Hi Marius,

Thank you for your reply!

I just quick checked your solution and added `"abort_on_first_run_crash": False` to `/usr/local/lib/python3.5/dist-packages/autosklearn/smbo.py` but I still got the same Exception...

My setup is very simple:
`autosklearn_model = autosklearn.classification.AutoSklearnClassifier(include_estimators=["sgd"], time_left_for_this_task=int(time_budget), per_run_time_limit=int(time_budget*1/5), output_folder=result_dir, tmp_folder=tmp_dir, delete_tmp_folder_after_terminate=False, delete_output_folder_after_terminate=False)`.
It is a multiclass task, and time_budget is 300 seconds.

**Comment by mlindauer** [comment]
hmmm... sorry, if this does not resolve your issue, I don't know how to fix it. Maybe Matthias will read this thread at some point and has an idea.

**Comment by mfeurer** [comment]
This is tricky to debug. Is it possible for you to add a print statement before the raise here https://github.com/automl/SMAC3/blob/master/smac/tae/execute_ta_run.py#L186 and post the error message?

@mlindauer we should think about improving the error message here...This is really not helpful.

**Comment by LishengSun** [comment]
Hi,

I added "going to raise FirstRunCrashedException" before the raise, and here is the complete error message:

`going to raise FirstRunCrashedException
Traceback (most recent call last):
  File "simulations_basic_models_and_autosklearn_tunning.py", line 496, in <module>
    autosklearn_model.fit(data.data['X_train'], data.data['Y_train'])
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 405, in fit
    dataset_name=dataset_name)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 304, in fit
    super(AutoSklearnEstimator, self).fit(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 23, in fit
    self._automl.fit(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 546, in fit
    return self._automl.fit(X, y, task, metric, feat_type, dataset_name)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/automl.py", line 195, in fit
    return self._fit(loaded_data_manager, metric)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/automl.py", line 449, in _fit
    _proc_smac.run_smbo()
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/smbo.py", line 662, in run_smbo
    smac.solver.incumbent = smac.solver.initial_design.run()
  File "/usr/local/lib/python3.5/dist-packages/smac/initial_design/single_config_initial_design.py", line 80, in run
    instance_specific=self.scenario.instance_specific.get(rand_inst, "0"))
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/evaluation/__init__.py", line 144, in start
    capped=capped)
  File "/usr/local/lib/python3.5/dist-packages/smac/tae/execute_ta_run.py", line 151, in start
    raise FirstRunCrashedException("First run crashed, abort. (To "
smac.tae.execute_ta_run.FirstRunCrashedException: First run crashed, abort. (To prevent this, toggle the 'abort_on_first_run_crash'-option!)
`

**Comment by mfeurer** [comment]
Sorry for my unhelpful request to add *a* print statement. Could you please add the following print statement?

    print(status, cost, runtime, additional_info)

in line 148?

**Comment by LishengSun** [comment]
Line 148? OK. I added it: 
`
if self.stats.is_budget_exhausted():
    print(status, cost, runtime, additional_info)
    raise BudgetExhaustedException(
        "Skip target algorithm run due to exhausted "
        "configuration budget")
` 

But I don't get more information than before, I still get the same exception:
`[ERROR] [2017-08-29 04:25:22,087:root] Traceback (most recent call last):
  File "DEBsimulations_basic_models_and_autosklearn_tunning.py", line 518, in <module>
    autosklearn_model.fit(data.data['X_train'], data.data['Y_train'])
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 405, in fit
    dataset_name=dataset_name)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 304, in fit
    super(AutoSklearnEstimator, self).fit(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 23, in fit
    self._automl.fit(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/estimators.py", line 546, in fit
    return self._automl.fit(X, y, task, metric, feat_type, dataset_name)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/automl.py", line 195, in fit
    return self._fit(loaded_data_manager, metric)
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/automl.py", line 449, in _fit
    _proc_smac.run_smbo()
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/smbo.py", line 662, in run_smbo
    smac.solver.incumbent = smac.solver.initial_design.run()
  File "/usr/local/lib/python3.5/dist-packages/smac/initial_design/single_config_initial_design.py", line 80, in run
    instance_specific=self.scenario.instance_specific.get(rand_inst, "0"))
  File "/usr/local/lib/python3.5/dist-packages/autosklearn/evaluation/__init__.py", line 144, in start
    capped=capped)
  File "/usr/local/lib/python3.5/dist-packages/smac/tae/execute_ta_run.py", line 181, in start
    raise FirstRunCrashedException("First run crashed, abort. (To "
smac.tae.execute_ta_run.FirstRunCrashedException: First run crashed, abort. (To prevent this, toggle the 'abort_on_first_run_crash'-option!)`

**Comment by mfeurer** [comment]
Sorry, it seems that your not using the SMAC version I assumed. Could you please add the statement [here](https://github.com/automl/SMAC3/blob/v0.5.0/smac/tae/execute_ta_run.py#L148)?

**Comment by LishengSun** [comment]
You are right. We don't have the same execute_ta_run.py. Actually I am using the docker image (https://hub.docker.com/r/mfeurer/auto-sklearn/tags/).

**Comment by LishengSun** [comment]
And actually I got this exception when trying to use sgd for multilabel and multiclass classification problem in AutoML2016, and when I break them down to multiple binary classifications, I no longer get this error. But I think it's still better to be able to turn off the 'abort_on_first_run_crash'-option, and get more informative error messages.

**Comment by mfeurer** [comment]
Yes, definitively. I will follow up on this when I'm back in Germany.

**Comment by falfaroalmagro** [comment]
Hi all,

I am also having a FirstRunCrashedException problem. The dataset I am using is OK (I have tried it with a couple of easy classifiers without any problem). 

Any chance to solve this bug soon?

Would I help if I revert to a previous version of auto-sklearn?

Cheers.

**Comment by mfeurer** [comment]
Hey, I'm currently out of office until the end of September. I'll have a look when I'm back. For now to debug this, it would be great if you can put a print statement as mentioned above.

**Comment by JosPolfliet** [comment]
Same here
```
import numpy as np
import autosklearn.regression

cls = autosklearn.regression.AutoSklearnRegressor()
x_train = np.random.normal(size=(2, 100))
y_train = 2*x_train + np.random.normal(size=(1, 100))
cls.fit(x_train, y_train)
```
gives
```
[WARNING] [2017-10-04 18:11:53,119:AutoMLSMBO(1)::1bf8e11a974776bf7db4c3fa37498c49] Could not find meta-data directory /Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/metalearning/files/r2_regression_dense
[WARNING] [2017-10-04 18:11:53,145:EnsembleBuilder(1):1bf8e11a974776bf7db4c3fa37498c49] Error loading predictions_ensemble_1_00001.npy: <class 'ValueError'> - continuous-multioutput
[ERROR] [2017-10-04 18:11:53,152:EnsembleBuilder(1):1bf8e11a974776bf7db4c3fa37498c49] Caught ValueError: continuous-multioutput
---------------------------------------------------------------------------
FirstRunCrashedException                  Traceback (most recent call last)
<ipython-input-4-9866da71eca1> in <module>()
      5 x_train = np.random.normal(size=(2, 100))
      6 y_train = 2*x_train + np.random.normal(size=(1, 100))
----> 7 cls.fit(x_train, y_train)

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/estimators.py in fit(self, X, y, metric, feat_type, dataset_name)
    487         return super(AutoSklearnRegressor, self).fit(X=X, y=y, metric=metric,
    488                                                      feat_type=feat_type,
--> 489                                                      dataset_name=dataset_name)
    490 
    491     def predict(self, X, batch_size=None, n_jobs=1):

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/estimators.py in fit(self, *args, **kwargs)
    302     def fit(self, *args, **kwargs):
    303         self._automl = self.build_automl()
--> 304         super(AutoSklearnEstimator, self).fit(*args, **kwargs)
    305 
    306     def fit_ensemble(self, y, task=None, metric=None, precision='32',

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/estimators.py in fit(self, *args, **kwargs)
     21 
     22     def fit(self, *args, **kwargs):
---> 23         self._automl.fit(*args, **kwargs)
     24 
     25     def refit(self, X, y):

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/estimators.py in fit(self, X, y, metric, feat_type, dataset_name)
    619             metric = r2
    620         return self._automl.fit(X=X, y=y, task=REGRESSION, metric=metric,
--> 621                                 feat_type=feat_type, dataset_name=dataset_name)

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/automl.py in fit(self, X, y, task, metric, feat_type, dataset_name)
    193                                             dataset_name=dataset_name)
    194 
--> 195         return self._fit(loaded_data_manager, metric)
    196 
    197     def fit_automl_dataset(self, dataset, metric):

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/automl.py in _fit(self, datamanager, metric)
    447                                     configuration_mode=self._configuration_mode)
    448             self.runhistory_, self.trajectory_, self.fANOVA_input_ = \
--> 449                 _proc_smac.run_smbo()
    450             trajectory_filename = os.path.join(
    451                 self._backend.get_smac_output_directory(self._seed) + '_run1',

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/smbo.py in run_smbo(self)
    659         smac.solver.stats.start_timing()
    660         # == first, evaluate all metelearning and default configurations
--> 661         smac.solver.incumbent = smac.solver.initial_design.run()
    662 
    663         for challenger in metalearning_configurations:

/Users/jos/anaconda/lib/python3.6/site-packages/smac/initial_design/single_config_initial_design.py in run(self)
     78             cutoff=self.scenario.cutoff,
     79             seed=initial_seed,
---> 80             instance_specific=self.scenario.instance_specific.get(rand_inst, "0"))
     81 
     82         self.stats.inc_changed += 1  # first incumbent

/Users/jos/anaconda/lib/python3.6/site-packages/autosklearn/evaluation/__init__.py in start(self, config, instance, cutoff, seed, instance_specific, capped)
    142         return super().start(config=config, instance=instance, cutoff=cutoff,
    143                              seed=seed, instance_specific=instance_specific,
--> 144                              capped=capped)
    145 
    146     def run(self, config, instance=None,

/Users/jos/anaconda/lib/python3.6/site-packages/smac/tae/execute_ta_run.py in start(self, config, instance, cutoff, seed, instance_specific, capped)
    148 
    149         if self.stats.ta_runs == 0 and status == StatusType.CRASHED:
--> 150             raise FirstRunCrashedException("First run crashed, abort. (To "
    151                                            "prevent this, toggle the "
    152                                            "'abort_on_first_run_crash'"

FirstRunCrashedException: First run crashed, abort. (To prevent this, toggle the 'abort_on_first_run_crash'-option!)
```

Output of pip freeze:
```
adal==0.4.5
alabaster==0.7.9
alembic==0.9.3
anaconda-client==1.6.0
anaconda-navigator==1.5.2
anaconda-project==0.4.1
appnope==0.1.0
appscript==1.0.1
astroid==1.4.9
astropy==2.0.1
auto-sklearn==0.2.0
awscli==1.11.117
azure-common==1.1.6
azure-datalake-store==0.0.11
azure-mgmt-datalake-nspkg==2.0.0
azure-mgmt-datalake-store==0.1.5
azure-mgmt-nspkg==2.0.0
azure-mgmt-resource==1.1.0
azure-nspkg==2.0.0
Babel==2.3.4
backports.shutil-get-terminal-size==1.0.0
bcolz==0.12.1
beautifulsoup4==4.5.3
bintrees==2.0.7
bitarray==0.8.1
bitcoin==1.1.42
blaze==0.10.1
bokeh==0.12.4
boto==2.45.0
boto3==1.4.4
botocore==1.5.80
Bottleneck==1.2.1
cachetools==2.0.0
certifi==2017.4.17
cffi==1.9.1
chardet==2.3.0
chest==0.2.3
click==6.7
cloudpickle==0.2.2
clyent==1.2.2
CodernityDB==0.5.0
cognitive-luis==0.1
coinbase==2.0.6
coinmarketcap==3.0.1
colorama==0.3.7
conda==4.3.27
conda-build==2.1.10
conda-verify==2.0.0
configobj==5.0.6
ConfigSpace==0.3.10
contextlib2==0.5.4
couchbase==2.2.5
cryptography==1.7.1
cycler==0.10.0
cymem==1.31.2
cyordereddict==1.0.0
Cython==0.25.2
cytoolz==0.8.2
dask==0.15.0
datashape==0.5.4
decorator==4.0.11
devp2p==0.9.2
dill==0.2.5
docutils==0.13.1
empyrical==0.3.0
et-xmlfile==1.0.1
ethereum==2.0.4
fastcache==1.0.2
filelock==2.0.7
Flask==0.12
Flask-Cors==3.0.2
future==0.16.0
gapic-google-cloud-datastore-v1==0.15.3
gdax==1.0.6
gevent==1.2.1
gipc==0.6.0
glueviz==0.9.1
google-auth==1.0.1
google-auth-httplib2==0.0.2
google-cloud-core==0.25.0
google-cloud-datastore==1.1.0
google-gax==0.15.13
googleapis-common-protos==1.5.2
greenlet==0.4.11
grpcio==1.4.0
h5py==2.7.0
HeapDict==1.0.0
httplib2==0.10.3
idna==2.2
imagesize==0.7.1
inflection==0.3.1
intervaltree==2.1.0
ipykernel==4.5.2
ipython==5.1.0
ipython-genutils==0.1.0
ipywidgets==5.2.2
isodate==0.5.4
isort==4.2.5
itsdangerous==0.24
jdcal==1.3
jedi==0.9.0
Jinja2==2.9.4
jmespath==0.9.3
joblib==0.11
jsonschema==2.5.1
jupyter==1.0.0
jupyter-client==4.4.0
jupyter-console==5.0.0
jupyter-core==4.2.1
Keras==2.0.6
keyring==10.3.3
lazy-object-proxy==1.2.2
leveldb==0.194
liac-arff==2.1.1
llvmlite==0.19.0
lmdb==0.93
locket==0.2.0
lockfile==0.12.2
Logbook==1.1.0
lru-dict==1.1.6
lxml==3.7.2
Mako==1.0.7
MarkupSafe==0.23
matplotlib==2.0.2
mistune==0.7.3
more-itertools==3.2.0
mpmath==0.19
msrest==0.4.10
msrestazure==0.4.8
multipledispatch==0.4.9
murmurhash==0.26.4
nbconvert==4.2.0
nbformat==4.2.0
networkx==1.11
nltk==3.2.2
nose==1.3.7
notebook==4.3.1
numba==0.34.0
numexpr==2.6.2
numpy==1.12.1
numpydoc==0.6.0
oauth2client==3.0.0
oauthlib==2.0.2
odo==0.5.0
olefile==0.44
openpyxl==2.4.1
pandas==0.20.3
pandas-datareader==0.5.0
pandas-profiling==1.4.0
partd==0.3.8
pathlib==1.0.1
pathlib2==2.2.0
patsy==0.4.1
pbkdf2==1.3
pep8==1.7.0
pexpect==4.2.1
pickleshare==0.7.4
Pillow==4.0.0
pkginfo==1.4.1
plac==0.9.6
ply==3.10
preshed==0.46.4
prompt-toolkit==1.0.9
proto-google-cloud-datastore-v1==0.90.4
protobuf==3.3.0
psutil==5.0.1
ptyprocess==0.5.1
py==1.4.32
pyasn1==0.1.9
pyasn1-modules==0.0.9
pycosat==0.6.1
pycparser==2.17
pycrypto==2.6.1
pycryptodome==3.4.6
pycurl==7.43.0
pyelliptic==1.5.7
pyethapp==1.5.0
pyethash==0.1.27
pyflakes==1.5.0
Pygments==2.1.3
PyJWT==1.5.0
pylint==1.6.4
pynisher==0.4.2
pyOpenSSL==16.2.0
pyparsing==2.1.4
pyrfr==0.4.0
pysha3==1.0.2
pytest==3.0.5
python-crfsuite==0.9.2
python-dateutil==2.6.0
python-editor==1.0.3
pytz==2016.10
PyWavelets==0.5.2
PyYAML==3.12
pyzmq==16.0.2
QtAwesome==0.4.3
qtconsole==4.2.1
QtPy==1.2.1
Quandl==3.1.0
rasa-nlu==0.8.11
redis==2.10.5
repoze.lru==0.6
requests==2.13.0
requests-file==1.4.2
requests-ftp==0.3.1
requests-oauthlib==0.8.0
rlp==0.6.0
rope-py3k==0.9.4.post1
rsa==3.4.2
s3transfer==0.1.10
scikit-image==0.13.0
scikit-learn==0.18.1
scipy==0.19.1
scrypt==0.8.0
seaborn==0.7.1
semver==2.7.7
simplegeneric==0.8.1
singledispatch==3.4.0.3
six==1.10.0
smac==0.5.0
snowballstemmer==1.2.1
sockjs-tornado==1.0.3
sortedcontainers==1.5.7
Sphinx==1.5.1
sputnik==0.9.3
spyder==3.1.2
SQLAlchemy==1.1.5
statistics==1.0.3.5
statsmodels==0.8.0
sympy==1.0
tables==3.4.2
tensorflow==1.1.0
terminado==0.6
Theano==0.9.0
thinc==5.0.8
tinyrpc==0.5
toolz==0.8.2
tornado==4.4.2
traitlets==4.3.1
typing==3.6.1
ujson==1.35
unicodecsv==0.14.1
visdom==0.1.5
wcwidth==0.1.7
websocket-client==0.40.0
Werkzeug==0.11.15
widgetsnbextension==1.2.6
wrapt==1.10.8
xlrd==1.0.0
XlsxWriter==0.9.6
xlwings==0.10.2
xlwt==1.2.0
zipline==1.1.1
```

**Comment by mfeurer** [comment]
Will be fixed in the next release via #366 and 336823128dfb1c6443100fd542550629711eaff7. Thank you very much for reporting this issue.
